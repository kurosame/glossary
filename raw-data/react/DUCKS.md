## category

react

## titles

Ducks
Redux

## description

Redux のデザインパターン  
Action や Reducer を modules にパッケージする

vuejs-boilerplate でも使ってる気がする  
1 ファイルにはしてないけど、actions, getters, mutations, state を import して modules でパッケージして export してる  
そして Store 作るときに modules をセットしている
