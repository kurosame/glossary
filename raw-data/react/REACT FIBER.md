## category

react

## titles

React Fiber
react-fiber

## description

Begin, Complete, Commit のフェーズがある

### Begin, Complete

Component 単位で実行される  
componentWillMount, componentWillReceiveProps, shouldComponentUpdate, componentWillUpdate  
非同期レンダリングでは中断・再開により、複数回ライフサイクル関数が呼ばれる可能性がある

### Commit

全ての Fiber が完了した時に実行される  
DOM 更新などの副作用が発生する処理はここでまとめて反映される  
componentDidMount, componentDidUpdate, componentWillUnmount

### Time Slicing

Fiber により画面のレンダリングが非同期になると、Fiber 単位で処理の優先度を付けることができる  
例えばユーザ操作の優先度を高くすることで、画面がレンダリング中でもユーザ操作を優先して処理することが可能となる

### Suspense

非同期通信をしている間、画面のレンダリングを中断できる  
非同期通信が完了した後に、画面をレンダリングすることが可能  
render に Promise を投げて、その Promise が完了した後に render するみたいな？
