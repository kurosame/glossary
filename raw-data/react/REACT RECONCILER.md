## category

react

## titles

React Reconciler
react-reconciler

## description

VDOM と呼ばれている部分

1. ReactElement のツリーを構築
1. Component をマウント、アンマウント、更新
1. ライフサイクル関数を呼出す
1. 現在の DOM との差分を計算する
1. renderer に差分を伝える
