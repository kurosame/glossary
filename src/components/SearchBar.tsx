import { Icon, Input, Paper } from '@material-ui/core'
import { Search } from '@material-ui/icons'
import React from 'react'
import styled from 'styled-components'

interface Props {
  onSearch: React.ChangeEventHandler<HTMLInputElement>
}

const StyledPaper = styled(Paper)`
  display: flex;
  align-items: center;
  width: 300px;
`
const StyledInput = styled(Input)`
  flex-grow: 1;
  margin-left: 10px;
`
const StyledIcon = styled(Icon)`
  margin: 10px;
`

const SearchBar: React.FC<Props> = p => (
  <StyledPaper>
    <StyledInput placeholder="Search..." onChange={p.onSearch} inputProps={{ 'data-testid': 'search-bar-input' }} />
    <StyledIcon>
      <Search />
    </StyledIcon>
  </StyledPaper>
)

export default SearchBar
