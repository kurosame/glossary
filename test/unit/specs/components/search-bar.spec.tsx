/**
 * @jest-environment jsdom
 */

import React from 'react'
import { render, fireEvent, cleanup, RenderResult } from '@testing-library/react'
import SearchBar from '@/components/SearchBar'

let onSearch: jest.Mock
let wrapper: RenderResult
beforeEach(() => {
  onSearch = jest.fn()
  wrapper = render(<SearchBar onSearch={onSearch} />)
})
afterEach(() => {
  cleanup()
  jest.restoreAllMocks()
})

test('Run `onChange`', () => {
  fireEvent.change(wrapper.getByTestId('search-bar-input'), {
    target: { value: 'test' }
  })

  expect(onSearch).toBeCalled()
})

test('Match the snapshot', () => {
  expect(wrapper.asFragment()).toMatchSnapshot()
})
