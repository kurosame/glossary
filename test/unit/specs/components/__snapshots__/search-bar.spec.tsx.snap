// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-11l9k5w-MuiPaper-root"
  >
    <div
      class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary css-168pjvi-MuiInputBase-root-MuiInput-root"
    >
      <input
        class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        data-testid="search-bar-input"
        placeholder="Search..."
        type="text"
        value=""
      />
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1ddhjnx-MuiSvgIcon-root"
      data-testid="SearchIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
      />
    </svg>
  </div>
</DocumentFragment>
`;
